org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/MarkerUpdater.java

```diff
@@ -0,0 +1,141 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarlint.eclipse.core.internal.tracking;
+
+import java.util.Collection;
+import java.util.Locale;
+import org.eclipse.core.filebuffers.FileBuffers;
+import org.eclipse.core.filebuffers.ITextFileBuffer;
+import org.eclipse.core.filebuffers.ITextFileBufferManager;
+import org.eclipse.core.filebuffers.LocationKind;
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IMarker;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.resources.IWorkspace;
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.core.runtime.Path;
+import org.eclipse.jface.text.IDocument;
+import org.sonarlint.eclipse.core.internal.PreferencesUtils;
+import org.sonarlint.eclipse.core.internal.SonarLintCorePlugin;
+import org.sonarlint.eclipse.core.internal.markers.FlatTextRange;
+import org.sonarlint.eclipse.core.internal.markers.MarkerUtils;
+
+public class MarkerUpdater implements TrackingChangeListener {
+
+  private final ModulePathManager modulePathManager;
+
+  public MarkerUpdater(ModulePathManager modulePathManager) {
+    this.modulePathManager = modulePathManager;
+  }
+
+  @Override
+  public void onTrackingChange(String moduleKey, String relativePath, Collection<? extends Trackable> issues) {
+    String absolutePath = modulePathManager.getFilePath(moduleKey, relativePath);
+
+    IWorkspace workspace = ResourcesPlugin.getWorkspace();
+    IPath location = Path.fromOSString(absolutePath);
+    IFile file = workspace.getRoot().getFileForLocation(location);
+
+    try {
+      file.deleteMarkers(SonarLintCorePlugin.MARKER_ID, true, IResource.DEPTH_INFINITE);
+    } catch (CoreException e) {
+      SonarLintCorePlugin.getDefault().error(e.getMessage(), e);
+      return;
+    }
+
+    ITextFileBufferManager textFileBufferManager = FileBuffers.getTextFileBufferManager();
+    if (textFileBufferManager == null) {
+      return;
+    }
+
+    IPath path = file.getFullPath();
+    try {
+      textFileBufferManager.connect(path, LocationKind.IFILE, new NullProgressMonitor());
+      ITextFileBuffer textFileBuffer = textFileBufferManager.getTextFileBuffer(path, LocationKind.IFILE);
+      IDocument document = textFileBuffer.getDocument();
+
+      if (document != null) {
+        for (Trackable issue : issues) {
+          if (!issue.isResolved()) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS134)